reps/2022-12-14-native-pipelining-data.md

```diff
@@ -0,0 +1,259 @@
+
+Thanks to the flexibility of Ray, this change is not as large as it may seem. For example, Ray's lineage fault tolerance will continue to "just work" even in the pipelined execution model. The main challenge is defining a clear set of internal interfaces to make the migration incremental.
+
+#### Corner Cases
+
+There are a few APIs (i.e., the variations of .split()) that are challenging to pipeline. This is because for some datasources (e.g., csv, json), we do not know ahead of time how many records are in each file, hence we need to fully execute read tasks to perform exact splits.
+
+The short term solution is to treat split() as a special case, preserving backwards compatibility and logging a warning to the user that this triggers synchronous execution for certain datasources. We can add a new iter_split() API that does support pipelining in all use cases (but that has slightly different semantics from split()).
+
+### Interfaces
+
+We introduce a new interface between the operator and execution layers of Datasets, which enables flexible selection of the execution strategy (bulk vs pipelined). The operator layer defines "what" to execute (e.g., Arrow table operations), and the executor defines "how" it is executed (e.g., bulk vs pipelined). Previously, operator definition and execution were intertwined in the code.
+
+<img src="https://user-images.githubusercontent.com/14922/207695244-b5a90d67-bdc9-4519-9bf1-fa01ecb877a4.png" alt="drawing" width="600px"/>
+
+For example, suppose you are running the following "Two-stage inference example":
+
+```python
+	ray.data.read_parquet()
+		.map_batches(fn_1)
+		.map_batches(fn_2)
+		.map_batches(infer, compute="actors", num_gpus=1)
+		.write_parquet()
+```
+
+Datasets translates the user API calls into an optimized series of operators, each of which should be run in a separate task:
+
+```
+	Read() -> Map([read, fn_1, fn_2]) -> Map(infer, GPU) -> Map(write)
+```
+
+The Executor is responsible for deciding how to execute the final operator DAG. It could execute the operators one at a time (bulk synchronous execution), or concurrently (pipelined execution). In the above example, the pipelined strategy would use much less peak memory, since it avoids fully materializing the data between the CPU and GPU stages.
+
+#### Class Hierarchy
+
+The proposed interface class hierarchy is as follows:
+- **RefBundle**: A group of Ray object references and their metadata. This is a logical "Block" in the Dataset.
+- **PhysicalOperator**: Node in the operator DAG that transforms one or more RefBundle input streams to a RefBundle output stream.
+    * Example subclasses (these may be grouped into sub-abstractions such as OneToOneOperator and AllToAllOperator; the definition of these abstractions is outside the scope of this REP):
+        + Read/Write
+        + Map
+        + Sort
+        + RandomShuffle
+        + ...
```

Comment:
would windowed all-to-all operations be supported here?

For example, a random shuffle, but it doesn't need to operate globally and it's sufficient to shuffle within a certain window size.

Reply:
Yes, we should be able to support this. We could add a special `Window` physical operator that buffers groups of blocks and releases them. We would also have to extend the operator interface slightly to introduce the notion of resetting the state for each window.